proto/struct.go

```diff
@@ -0,0 +1,387 @@
+
+		for _, f := range unique {
+			size := f.codec.size(f.pointer(p), f.makeFlags(flags))
+			if size > 0 {
+				n += int(f.tagsize) + size
+				if f.embedded() {
+					n += sizeOfVarint(uint64(size))
+				}
+				flags = flags.without(wantzero)
+			}
+		}
+
+		for _, f := range repeated {
+			size := f.codec.size(f.pointer(p), f.makeFlags(flags))
+			if size > 0 {
+				n += size
+				flags = flags.without(wantzero)
+			}
+		}
+
+		return n
+	}
+}
+
+func structEncodeFuncOf(t reflect.Type, fields []structField) encodeFunc {
+	var inlined = inlined(t)
+	var unique, repeated []*structField
+
+	for i := range fields {
+		f := &fields[i]
+		if f.repeated() {
+			repeated = append(repeated, f)
+		} else {
+			unique = append(unique, f)
+		}
+	}
+
+	return func(b []byte, p unsafe.Pointer, flags flags) (int, error) {
+		if p == nil {
+			return 0, nil
+		}
+
+		if !inlined {
+			flags = flags.without(inline | toplevel)
+		} else {
+			flags = flags.without(toplevel)
+		}
+		offset := 0
+
+		for _, f := range unique {
+			fieldFlags := f.makeFlags(flags)
+			elem := f.pointer(p)
+			size := f.codec.size(elem, fieldFlags)
+
+			if size > 0 {
+				n, err := encodeTag(b[offset:], f.fieldNumber(), f.wireType())
+				offset += n
+				if err != nil {
+					return offset, err
+				}
+
+				if f.embedded() {
+					n, err := encodeVarint(b[offset:], uint64(size))
+					offset += n
+					if err != nil {
+						return offset, err
+					}
+				}
+
+				if (len(b) - offset) < size {
+					return len(b), io.ErrShortBuffer
+				}
+
+				n, err = f.codec.encode(b[offset:offset+size], elem, fieldFlags)
+				offset += n
+				if err != nil {
+					return offset, err
+				}
+
+				flags = flags.without(wantzero)
+			}
+		}
+
+		for _, f := range repeated {
+			n, err := f.codec.encode(b[offset:], f.pointer(p), f.makeFlags(flags))
+			offset += n
+			if err != nil {
+				return offset, err
+			}
+			if n > 0 {
+				flags = flags.without(wantzero)
+			}
+		}
+
+		return offset, nil
+	}
+}
+
+func structDecodeFuncOf(t reflect.Type, fields []structField) decodeFunc {
+	maxFieldNumber := fieldNumber(0)
+
+	for _, f := range fields {
+		if n := f.fieldNumber(); n > maxFieldNumber {
+			maxFieldNumber = n
+		}
+	}
+
+	fieldIndex := make([]*structField, maxFieldNumber+1)
+
+	for i := range fields {
+		f := &fields[i]
+		fieldIndex[f.fieldNumber()] = f
+	}
+
+	return func(b []byte, p unsafe.Pointer, flags flags) (int, error) {
+		flags = flags.without(toplevel)
+		offset := 0
+
+		for offset < len(b) {
+			fieldNumber, wireType, n, err := decodeTag(b[offset:])
+			offset += n
+			if err != nil {
+				return offset, err
+			}
+
+			i := int(fieldNumber)
+			f := (*structField)(nil)
+
+			if i >= 0 && i < len(fieldIndex) {
+				f = fieldIndex[i]
+			}
+
+			if f == nil {
+				skip := 0
+				size := uint64(0)
+				switch wireType {
+				case varint:
+					_, skip, err = decodeVarint(b[offset:])
+				case varlen:
+					size, skip, err = decodeVarint(b[offset:])
+					skip += int(size)
```

Comment:
Another case where it might overflow.

Reply:
I wish there was a static checker for these 😓